### PR TITLE
Address further feedback from reviewers

### LINKED ISSUE #59: Review feedback to address
**Meta-reviewer**

- [ ] The most important one is that the blog post would be a lot stronger if it included an example from the machine learning literature that requires materialization of the Jacobian or Hessian matrix rather then access to these objects via vector products. In other words, please tell the ICLR audience why they should care about this post and take the time to read it. [...] Please give an example from machine learning that could take advantage of the machinery described in this blog post.
- [x] The less important one is the remark that the "...by people more familiar with low-level programming languages" comment is odd. I agree: better to simply say that this theory has been developed outside the ML community and leave it at that.
- [x] The caption of Figure 16 specifies the wrong f(x). Please fix this error.

**Reviewer 1**

- [x] ~~For example, reverse mode AD was invented as early as 1970, and before backprop was invented, even though the latter is a special case of the former. So discussing the history and connecting the two together, and explaining the difference, would be super interesting.~~
- [ ] Also, as the blog mentioned, in deep learning, typically there is only one loss function, so no Jacobi, so it would be nice to understand how the techniques in this paper could benefit the ML community. [...] ~~Adding the connection between the technique discussed here and ML would be super interesting. [...] History and connection with ML would make it stronger~~
- [x] "One reason is that the underlying theory was developed outside of the ML research ecosystem, by people more familiar with low-level programming languages." This seems a bit odd to say. ML people used to work in low-level programming languages as well. My understanding is that AD was invented for optimization by people who are numerical analysis researchers
- [x] " We consider a convolutional filter of size 5×5... an input size of 28x28x1...576×784 Jacobian" Would be helpful to explain where these numbers like 576 comes from.$

**Reviewer 2**

- [x] On a subjective note, a demonstration in Torch or JAX (both have 2nd order derivatives AFAIK) would have a greater impact over Julia, as more of the ML community uses these tools for real world applications and cases .
- [x] Consider replacing impossible with intractable when needed
- [x] Linear maps are formally defined (after figure 2), but a novice reader might miss their property, e.g. why are they merely an approximation for the Jacobian? -or- in what sense are the dashed matrices differ from the solid ones
- [x] Similar to the previous point, it is stated that materialization does not materialize any intermediate Jacobians, consider adding another sentence on when and especially why materialization differ.
- [x] Figure 6 - consider reminding the black-box here again - the numbers inside were not clear (to me) until a second read.
- [x] ~~Figure 13 - consider making this more general, each row has exactly two indexes :-)~~

### PR DESCRIPTION
**Pull request by adrhill**
WIP to close #59.

### CODE REVIEW

File: _posts/2025-04-28-sparse-autodiff.md [hunk 1]
```diff
@@ -755,8 +755,11 @@ We complement this tutorial with a demonstration of automatic sparse differentia
 While still at an early stage of development, we hope that such an example of unified pipeline for sparse Jacobians and Hessians can inspire developers in other languages to revisit ASD.
 
 <aside class="l-body box-note" markdown="1">
-The authors of this blog post are all developers of the ASD ecosystem in Julia. We are not aware of a similar ecosystem in Python or R, which is why we chose Julia to present it.
-The closest counterpart we know is coded in C, namely the combination of ADOL-C <d-cite key="waltherGettingStartedADOLC2009"></d-cite> and ColPack <d-cite key="gebremedhinColPackSoftwareGraph2013"></d-cite>.
+The authors of this blog post are all developers of the ASD ecosystem in Julia.
+We use Julia for our demonstration since we are not aware of a similar ecosystem in Python or R.
+At the time of writing, PyTorch, TensorFlow, and JAX lack sparsity detection and coloring capabilities.
```
Comment (gdalle) [inline _posts/2025-04-28-sparse-autodiff.md:760]:
```suggestion
At the time of writing, PyTorch, TensorFlow, and JAX lack comparable sparsity detection and coloring capabilities.
We are only aware of one JAX library for sparse differentiation, [`sparsejac`](https://github.com/mfschubert/sparsejac), which does not support sparsity detection, uses less efficient graph encodings and has no functionality for symmetric (Hessian) matrices. 
```

File: _posts/2025-04-28-sparse-autodiff.md [hunk 2]
```diff
@@ -749,6 +749,32 @@ This backup storage enables the use of **fewer distinct colors**, reducing the c
...
+
+The blog post goes on to argue that instead of materializing the full Hessian matrix, 
+matrix-free iterative solvers should be combined with Hessian operators by iteratively computing HVPs: 
+
+> When evaluating the term $$[\nabla^2f(\theta_k)]^{-1}\nabla f(\theta_k)$$, it would be very inefficient to first compute the full Hessian matrix $$\nabla^2f(\theta_k)$$, then invert it and finally multiply this with the gradient $$\nabla f(\theta_k)$$.
+
+While this holds true for dense Hessians, the opposite can be the case for sparse Hessians.
+Materializing a full Hessian matrix allows for an entirely new class of linear solvers to be used, e.g. enabling matrix factorization.
+Whether these are more performant depends on the number of colors in the sparity pattern of the Hessian 
+vs. the required numerical precision of the iterative solve.
```
Comment (amontoison) [inline _posts/2025-04-28-sparse-autodiff.md:776]: Iterative solvers are typically a last resort when memory limitations prevent the use of ASD.

File: _posts/2025-04-28-sparse-autodiff.md [hunk 3]
```diff
@@ -749,6 +749,32 @@ This backup storage enables the use of **fewer distinct colors**, reducing the c
 Powell and Toint <d-cite key="powellEstimationSparseHessian1979"></d-cite> were the first to notice symmetry-related optimizations, before Coleman and Moré <d-cite key="colemanEstimationSparseHessian1984"></d-cite> made the connection to graph coloring explicit.
 While symmetric coloring and decompression are more computationally expensive than their nonsymmetric counterparts, this cost is typically negligible compared to the savings we get from fewer HVPs.
 
+## Applications
+
+ASD is useful in applications which require the computation of full Jacobian or Hessian matrices with sparsity.
+One such example is given in the 2024 ICLR blog post *How to compute Hessian-vector products?* <d-cite key="dagreouHowComputeHessianvector2024"></d-cite>,
+where Hessians are used for second-order optimization:
+
+> When trying to find the minimum of the function $$f$$, methods that account for the second-order information often rely on the product between the inverse Hessian and a vector to find a good update direction.
```
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:758]: Can't change this, it's a quote:
https://iclr-blogposts.github.io/2024/blog/bench-hvp/#inverse-hessian-vector-products-ihvps-in-optimization

File: _posts/2025-04-28-sparse-autodiff.md [hunk 4]
```diff
@@ -749,6 +749,32 @@ This backup storage enables the use of **fewer distinct colors**, reducing the c
...
+> For instance, Newton's method relies on update rules of the form
+> 
+> \begin{equation}
+>   \notag
+>   \theta_{k+1} = \theta_k - \eta_k[\nabla^2f(\theta_k)]^{-1}\nabla f(\theta_k)
```
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:763]: Can't change this, it's a quote:
https://iclr-blogposts.github.io/2024/blog/bench-hvp/#inverse-hessian-vector-products-ihvps-in-optimization

File: _posts/2025-04-28-sparse-autodiff.md [hunk 5]
```diff
@@ -749,6 +749,32 @@ This backup storage enables the use of **fewer distinct colors**, reducing the c
...
+> \end{equation}
+> 
+> for some step-size $$\eta_k>0$$.
+
+The blog post goes on to argue that instead of materializing the full Hessian matrix, 
+matrix-free iterative solvers should be combined with Hessian operators by iteratively computing HVPs: 
+
+> When evaluating the term $$[\nabla^2f(\theta_k)]^{-1}\nabla f(\theta_k)$$, it would be very inefficient to first compute the full Hessian matrix $$\nabla^2f(\theta_k)$$, then invert it and finally multiply this with the gradient $$\nabla f(\theta_k)$$.
```
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:771]: Yes, that's what they claim.

File: _posts/2025-04-28-sparse-autodiff.md (hunk 2 above)
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:776]: So you're saying we can make even stronger claims here?

File: _posts/2025-04-28-sparse-autodiff.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:763]: Using our notation, we would capitalize the thetas, but I think it's (luckily) close enough.

File: _posts/2025-04-28-sparse-autodiff.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (amontoison) [reply _posts/2025-04-28-sparse-autodiff.md:758]: I don't think that we should quote them in that case.
The descent direction is the solution of a linear system Ax = b and what we want is A \ b which can be obtained with an iterative or direct solvers. A^{-1}b is one way to obtain A \ b but the most irrelevant one in practice.

[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:758]: Let's get @gdalle opinion on this.

File: _posts/2025-04-28-sparse-autodiff.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:771]: I think what they describe is what is commonly used in ML (?)

File: _posts/2025-04-28-sparse-autodiff.md (hunk 2 above)
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:776]: Could you elaborate on this in the blog post?

File: _posts/2025-04-28-sparse-autodiff.md [hunk 6]
```diff
@@ -749,6 +749,32 @@ This backup storage enables the use of **fewer distinct colors**, reducing the c
 Powell and Toint <d-cite key="powellEstimationSparseHessian1979"></d-cite> were the first to notice symmetry-related optimizations, before Coleman and Moré <d-cite key="colemanEstimationSparseHessian1984"></d-cite> made the connection to graph coloring explicit.
 While symmetric coloring and decompression are more computationally expensive than their nonsymmetric counterparts, this cost is typically negligible compared to the savings we get from fewer HVPs.
 
+## Applications
+
+ASD is useful in applications which require the computation of full Jacobian or Hessian matrices with sparsity.
```
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:754]: The original sentence was meant to emphasize that it's not useful for applications which just require gradients/VJP/JVP/HVP.

[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:754]: I agree with the original version

File: _posts/2025-04-28-sparse-autodiff.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:758]: In the machine learning literature, $A^{-1} b$ is often written to mean `A \ b` even though the system is not _actually_ solved by inverting the matrix. Leaving the quote is fine

File: _posts/2025-04-28-sparse-autodiff.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:771]: They don't use iterative solvers to compute $A^{-1}$, but to solve linear systems. As said above, $A^{-1}$ is a notation in this case, not a hint at an algorithm

File: _posts/2025-04-28-sparse-autodiff.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (amontoison) [reply _posts/2025-04-28-sparse-autodiff.md:763]: @adrhill You can quote the good version directly without A^{-1}
Screenshot_20250226_082918_Samsung Internet

File: _posts/2025-04-28-sparse-autodiff.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:771]: Fair enough

File: _posts/2025-04-28-sparse-autodiff.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:763]: Just to be clear: you're happy if I extend the quote to include the contents of the screenshot?

[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:763]: If so, addressed in [6c47b46](https://github.com/adrhill/iclr2025-sparse-autodiff/pull/60/commits/6c47b46ebee5436c1bae5dcbeb7e8f55019986b4).

File: _posts/2025-04-28-sparse-autodiff.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:771]: Hopefully this is addressed by extending the quote in [6c47b46](https://github.com/adrhill/iclr2025-sparse-autodiff/pull/60/commits/6c47b46ebee5436c1bae5dcbeb7e8f55019986b4).

File: _posts/2025-04-28-sparse-autodiff.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:758]: Hopefully addressed in [6c47b46](https://github.com/adrhill/iclr2025-sparse-autodiff/pull/60/commits/6c47b46ebee5436c1bae5dcbeb7e8f55019986b4).

File: _posts/2025-04-28-sparse-autodiff.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:771]: On the contrary, I agree with @amontoison and I think we should remove the quote. Essentially no one solves a linear system by computing the inverse, so it is not a goor argument for iterative solvers per se.

File: _posts/2025-04-28-sparse-autodiff.md [hunk 7]
```diff
@@ -749,14 +749,54 @@ This backup storage enables the use of **fewer distinct colors**, reducing the c
...
+> \end{equation}
+> 
+> with $$b = \nabla f(\theta_k)$$.
+> This approach is much more efficient as it avoids computing and storing the full Hessian matrix, and only computes the inverse of the matrix in the direction $$v$$.
+
+While this holds true for dense Hessian matrices, the opposite can be the case for sparse Hessians.
+By leveraging ASD, fewer HVPs and less memory are required to compute and store a sparse Hessian matrix.
+Materializing a full Hessian matrix allows the use of direct linear solvers, which are more robust than iterative solvers for ill-conditioned problems.
+Whether these are more performant depends on the number of colors in the sparsity pattern of the Hessian 
+vs. the required numerical precision of the iterative solve.
```
Comment (gdalle) [inline _posts/2025-04-28-sparse-autodiff.md:786]:
```suggestion
Whether these are more performant also depends on the number of colors in the sparsity pattern of the Hessian 
vs. the required numerical precision of the iterative solve.
But in general, iterative solves with linear operators are less widely supported than direct solves materialized matrices (e.g. in nonlinear optimization libraries).
```

File: _posts/2025-04-28-sparse-autodiff.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:771]: I wasn‘t sure what to make of this: https://github.com/adrhill/iclr2025-sparse-autodiff/pull/60#discussion_r1971938796

File: _posts/2025-04-28-sparse-autodiff.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:763]: No, I think the idea was to restrict the quote to this part:

> One computes the iHVP by solving the following linear system...

and then discuss the respective merits of direct and iterative solvers. Essentially remove mentions of $A^{-1}$

File: _posts/2025-04-28-sparse-autodiff.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:771]: Answered over there

File: _posts/2025-04-28-sparse-autodiff.md (hunk 2 above)
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md]: I suggest https://github.com/adrhill/iclr2025-sparse-autodiff/pull/60#discussion_r1976999620

File: _posts/2025-04-28-sparse-autodiff.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:763]: I’ll let Alexis remove it once he’s back from his talk. 👍

[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:763]: He's on the road for the next two weeks or so, I think we can take over

[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:763]: Feel free to delete whatever you want. As you mentioned before, $A^{-1}$ is simply common notation in ML and doesn't imply a matrix inverse.

[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md]: Done in the latest commit

File: _posts/2025-04-28-sparse-autodiff.md [hunk 8]
```diff
@@ -749,14 +749,34 @@ This backup storage enables the use of **fewer distinct colors**, reducing the c
...
+As the blog post argues, for such systems, matrix-free iterative solvers combine well with Hessian operators because they only require successive evaluations of HVPs.
+The main advantage of iterative solvers is their small memory footprint and their ease of implementation.
+
+However, when it is possible to materialize the whole Hessian matrix, direct linear solvers can be used (e.g. based on matrix factorization).
+As we have seen, ASD unlocks this option whenever the Hessian matrix is sparse enough.
+In terms of performance, computing a sparse Hessian matrix might require fewer HVPs overall than applying an iterative solver like the conjugate gradient method.
+The exact performance depends on the number of colors in the sparsity pattern of the Hessian, as well as the numerical precision expected from the iterative solver (which influences the number of iterations, hence the number of HVPs).
+In terms of numerical accuracy, direct solvers are more robust than their iterative counterparts.
+Finally, in terms of compatibility, some prominent nonlinear optimization libraries only support Hessian matrices, and not Hessian operators.
```
Comment (adrhill) [inline _posts/2025-04-28-sparse-autodiff.md:766]: This would be a stronger argument if we named (and cited) at least one,

Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:766]: Honestly I would put it almost at the top of the blog.
We can name the libraries @amontoison mentioned but we have to double-check in the docs first

Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:766]: We can also move the entire applications section, but I'm mostly referring to the last sentence here. 
Performance and numerical stability are more important than compatibility with some software most ML folks will never have heard of. 
Maybe less is more and we should just remove the last sentence entirely.

**Comment by adrhill** [comment]
Merging, revisions in follow-ups are of course welcome.
